src/Check/Producer.elm

```diff
@@ -343,3 +351,35 @@ map : (a -> b) -> Producer a -> Producer b
 map f prod =
     Producer (Random.map f prod.generator)
         Shrink.noShrink
+
+
+{-| Creates a Producer which, when asked to produce a value, will delegate
+to one of the Producers in the input list with equal probability.
+
+This method is useful for creating Producers for composite types. For example:
+
+```elm
+type Example =
+  Foo
+  | Bar Int
+  | Baz String
+
+exampleProducer : Producer Example
+exampleProducer =
+  oneOf [ const Foo
+        , int |> map Bar
+        , string |> map Baz
+        ]
+```
+-}
+oneOf : List (Producer a) -> Producer a
+oneOf ps =
+    let -- Producer a -> Generator (a, Shrinker a)
+        -- Given a `Producer a` get a Generator that produces pairs consisting of
+        -- a value and a shrinker to apply to that value
+        toGenAndShrinker p = p.generator |> Random.map (\a -> (a, p.shrinker))
+        -- Generator (a, Shrinker a)
+        generator = ps |> List.map toGenAndShrinker |> Random.Extra.choices
+        -- Shrinker (a, Shrinker a)
+        shrinker (a, subshrinker) = subshrinker a |> Lazy.List.map (\a' -> (a', subshrinker))
```

Comment:
When the Shrinker of the returned Producer receives a value we need to work out which Shrinker from the original list of producers we need to delegate to. This is why the constructed generator needs to return both the value and the shrinker of the producer that produced that value.